File: src/common/choice_cummulative.py
import random


def normalize(lst):
    '''
    input: un-normalize list
    output: normalize list


    print(normalize([1, 2, 3]))
    [0.16666666666666666, 0.3333333333333333, 0.5]
    print(normalize([0.16666666666666666, 0.3333333333333333, 0.5]))
    [0.16666666666666666, 0.3333333333333333, 0.5]
    '''
    total = float(sum(lst))
    lst = [i/total for i in lst]

    return lst


def cumulative(lst):
    """
    input  : int/float list
    output : Running total list

    print(cumulative([1, 2, 3]))
    # [1.0, 3.0, 6.0]
    print(cumulative([0.16666666666666666, 0.3333333333333333, 0.5]))
    # [0.16666666666666666, 0.5, 1.0]
    """
    output_lst = []
    running_total = 0.0
    for i in lst:
        running_total += i
        output_lst.append(running_total)

    return (output_lst)


def find_index_less_or_equal(float_lst, float_n):
    """
    print(find_index_less_or_equal([0.7, 0.8999999999999999, 0.9999999999999999], 0.9))
    #1
    print(find_index_less_or_equal([0.7, 0.8999999999999999, 0.9999999999999999], 0.8))
    #0
    """
    index = 0
    for i in range(len(float_lst)):
        item = float_lst[i]
        if float_n >= item:
            index = i+1
        else:
            break
    return index


def choice(lst, p, replace=False):
    """
    input: list
    output : selected 1 object from list based on p

    print(choice([1, 2, 3], [2, 5, 2], replace=False))
    #2
    #1,2,3 all possible based on probabilities.    
    """
    if replace != False:
        return None
    if len(p) != len(lst):
        return None

    normalized_p = normalize(p)
    normalized_cumulative_p = cumulative(normalized_p)
    random_float_n = random.random()  # float number between 0.0 to 1.0

    index = find_index_less_or_equal(normalized_cumulative_p, random_float_n)

    # print("""
    # lst                     : %s
    # p                       : %s
    # normalized_p            : %s
    # normalized_cumulative_p : %s
    # random_float            : %s
    # index                   : %s
    # """ % (lst, p, normalized_p, normalized_cumulative_p, random_float_n, index)
    #       )
    #
    # print(index)
    return lst[index]


def test_choice(lst=[1, 2, 3], p=[7, 2, 1], replace=False):
    """test choice function

    Args:
        lst (list, optional): input list. Defaults to [1, 2, 3].
        p (list, optional): input list elements probabilities. Defaults to [7, 2, 1].
        replace (bool, optional): [description]. Defaults to False.

    Returns:
        dict: return probabilities of getting each element over 1000 round
            example: {1: 711,2: 179, 3: 110 }
    """
    d = {}

    for i in range(1000):
        c = choice(lst, p, replace=False)
        if c not in d:
            d[c] = 1
        else:
            d[c] += 1
    return (d)


print(test_choice())

File: src/common/choice_cummulative.cpp
#include <iostream>
#include <vector>
#include <vector>
#include <numeric>
#include <random>

template <typename VT>
void printv(std::vector<VT> v)
{
    for (int i = 0; i < v.size(); i++)
    {
        std::cout << v.at(i) << " ";
    }
    std::cout << "\n";
}

// template <typename CV>
void convert_vec_to_float(std::vector<int> colsample_bytree_weight, int colsample_bytree_weight_factor, std::vector<float> &output)
{

    for (int i = 0; i < colsample_bytree_weight.size(); i++)
    {
        float fi = static_cast<float>(colsample_bytree_weight.at(i)) / static_cast<float>(colsample_bytree_weight_factor);
        output.push_back(fi);
    }
}

// template <typename N>
void normalize(std::vector<float> a, std::vector<float> &output)
{
    float sum_of_elems = 0;
    sum_of_elems = std::accumulate(a.begin(), a.end(),
                                   decltype(a)::value_type(0));
    for (int i = 0; i < a.size(); i++)
    {
        float item = a.at(i);
        float normalized = item / sum_of_elems;

        output.push_back(normalized);
    }
}

// template <typename CU>
void cumulative(std::vector<float> a, std::vector<float> &output)
{
    float running_total = 0;
    for (int i = 0; i < a.size(); i++)
    {
        running_total += a.at(i);
        output.push_back(running_total);
    }
}

// template <typename FI>
int find_index_less_or_equal(std::vector<float> a, float n)
{
    int index = 0;

    for (int i = 0; i < a.size(); i++)
    {
        float item = a.at(i);
        if (n >= item)
        {
            index = i + 1;
        }
        else
        {
            return index;
        }
    }
    return index;
}

template <typename TUU>
int choice_c(std::vector<TUU> input, std::vector<float> p)
{
    // std::cout << "\nChoice started\n";

    if (input.size() != p.size())
    {
        std::cout << "\ninput vector and probability vector size is not the same \n";
        return -1.0;
    }
    int conversion = 1; // will be used to divide by 1.0
    // std::vector<float> float_vector = {};
    std::vector<float> normalized = {};
    std::vector<float> cumulatived = {};

    // convert_vec_to_float(p, conversion, float_vector);
    normalize(p, normalized);
    // printv(normalized);

    cumulative(normalized, cumulatived);
    // printv(cumulatived);

    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_real_distribution<> dis(0, 1); //uniform distribution between 0 and 1
    float r = dis(gen);

    int index = find_index_less_or_equal(cumulatived, r);
    float selected;
    // selected = input.at(index);

    // std::cout << "\nChoice ended\n";
    return index;
}

// template <typename TUU>
// std::vector<TUU> choice(std::vector<TUU> a, int size, bool replace, std::vector<int> p)

template <typename TUU>
std::vector<TUU> choice_n(std::vector<TUU> input, int n, bool replace, std::vector<int> p)
{
    std::vector<float> pf = {};
    convert_vec_to_float(p, 1, pf);

    std::vector<TUU> output;
    for (int i = 0; i < n; i++)
    {

        int index = choice_c(input, pf);
        float item = input.at(index);
        output.push_back(item);
        input.erase(input.begin() + index);
        pf.erase(pf.begin() + index);
        //
    }
    return output;
}

// int main()
// {

//     std::vector<float> a = {1, 2, 3, 4, 5};
//     std::vector<int> p = {10, 5, 3, 2, 1};

//     // std::vector<int> a = {1, 2, 3, 4, 5};
//     // int b = 1;
//     // std::vector<float> c = {};
//     // std::vector<float> d = {};
//     // std::vector<float> e = {};
//     // convert_vec_to_float(a, b, c);
//     // printv(c);
//     // normalize(c, d);
//     // printv(d);
//     // cumulative(d, e);
//     // printv(e);
//     // int index = find_index_less_or_equal(e, 0.05);
//     // std::cout << index;

//     for (int i = 0; i <= 10; i++)
//     {
//         int n = 3;
//         printv(choice_n(a, n, false, p));
//         std::cout << '\n';
//     }

//     return 0;
// }
